server/chat/backend/agent/middleware/force_tool.py

```
@@ -2,32 +2,71 @@
 
-    def _tool_choice_for(self, model) -> object:
-        """Return the provider-native `tool_choice` value forcing a single tool.
+    @staticmethod
+    def _match_provider(module: str, name: str, llm_type: str) -> str | None:
+        for mod_key, cls_key, type_key, provider in _PROVIDER_SIGNATURES:
+            if mod_key in module or name == cls_key or type_key in llm_type:
+                return provider
+        return None
+
+    @staticmethod
+    def _infer_provider(model: Any) -> str | None:
```

Ruff (0.15.12)

[warning] Dynamically typed expressions (typing.Any) are disallowed in `model`

(ANN401)